<a href="https://colab.research.google.com/github/vkjadon/python/blob/main/opps.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###Class

A user-defined prototype for an object that defines a set of attributes that characterize any object is called the class. The attributes are data members (class variables and instance variables) and methods are functions to do a specific task and associated with a class.

The attributes and methods for an object of the class are accessed via dot notation.

Classes allow us to group our data called attributes and methods in a way that is easy to reuse and also easy to build upon when needed.

Let us create a class 'Employee' for a company as each individual employee is going to have specific attributes and methods.

In [ ]:
class Employee:
  pass

We can also understand a class as a blueprint for creating instances (objects). We can create employees using the class and these employees will be called as instance (object) of that class. Let us create two objects 'emp_1' and 'emp_2' of the Employee class by instantiating the class as below.

In [ ]:
emp_1=Employee()
emp_2=Employee()

The above are unique instances of the employee class.

###Object
A unique instance of a data structure that's defined by its class. An object comprises both data members (class variables and instance variables) and methods.

In [ ]:
print(emp_1)
print(emp_2)

###Instance Variable
Both of these are employee objects and both are unique as indicated by the unique memory locations. These instances ('emp_1' and 'emp_2') can have variables such as name, email and pay. These are called **instance variables**. Instance variables contain data that is unique to each instance.   
Let us set instance variables('firstName', 'lastName', 'email', 'pay') for both instances.

In [ ]:
emp_1.firstName='FirstName_1'
emp_1.lastName='LastName_1'
emp_1.email='FirstName.LastName@email.com'
emp_1.pay=50000

In [ ]:
emp_2.firstName='FirstName_2'
emp_2.lastName='LastName_2'
emp_2.email='FirstName_2.LastName_2@email.com'
emp_2.pay=40000

In [ ]:
print(emp_1.email)
print(emp_1)

FirstName.LastName@email.com


###Constructors
It is not a good practice to set the **instance variable** as above. We want that these variables are created automatically when an instance of a class is created. For this we use "\__init__\()" method in the class which is executed automatically whenever a **class is instantiated**. The \__init__\() method is also called as constructors. Constructors are generally used for **instantiating an object**. The task of constructors is to initialize(assign values) the instance of a class to the data members when an object of class is created. It stands for initializing and takes first argument as the instance itself and by convention it is called `self`. We can add other arguements after 'self'.    
So, we can modify the class as below

In [ ]:
class Employee:
  def __init__(self, firstName, lastName, email, pay):
    self.firstName=firstName
    self.lastName=lastName
    self.email=email
    self.pay=pay

In [ ]:
emp_1=Employee('FirstName_1', 'LastName_1','FirstName_1.LastName_1@email.com', 50000)
emp_2=Employee('FirstName_2', 'LastName_2','FirstName_2.LastName_2@email.com', 50000)

When we create these employees, the \__init__() method runs automatically and *self* take *emp_1* and *emp_2* when these are instantiated. The instance variables *firstName, lastName, email* and *pay* are the attribute of **Employee** class.

When we pass values to a method, these are called *arguements*. For examples we are passing 'FirstName_1', 'FirstName_2' etc to call __init__ method. These are arguements.

Parameters are the variables used for defining the methods. Here 'firstName', 'lastName' etc used to define __init__ method are the *Parameter*.

When we assign the *parameteres* to the *object variable*, these are called attributes.


In [ ]:
print(emp_1.email)
print(emp_2.email)

FirstName_1.LastName_1@email.com
FirstName_2.LastName_2@email.com


In [ ]:
print(Employee())

Let us try to print the full name. This we can do outside the class as below

In [ ]:
print(emp_1.firstName+' '+emp_1.lastName)

FirstName_1 LastName_1


Let us add some ability to perform some task in the class, eg. to print full name of an employee object. We add a method in the class itself to return full name of an employee.

In [ ]:
class Employee:
  def __init__(self, firstName, lastName, email, pay):
    self.firstName=firstName
    self.lastName=lastName
    self.email=email
    self.pay=pay

  def fullname(self):
    return emp_1.firstName+' '+ emp_1.lastName

In [ ]:
emp_1=Employee('FirstName_1', 'LastName_1','FirstName_1.LastName_1@email.com', 50000)
print(emp_1.fullname())

FirstName_1 LastName_1


When we call method `fullname()` on an instance, the `self` is automatically passed so, it is necessary to use `self` in the method declaration, else, it will throw the following error.

`TypeError: Employee.fullname() takes 0 positional arguments but 1 was given`

We can also run these methods using the class name. When we use class to call a method, we have to pass an instance as an argument.

In [ ]:
Employee.fullname(emp_1)

'FirstName_1 LastName_1'

###Class Variables and Instance Variables
A variable that is shared by all instances of a class is called **Class Variables**. The **class variables** are defined within a class but outside any of the methods. Class variables are not used as frequently as instance variables are. The class variable remain the same for all the instances. To undestand this, let us add another method to increase the salary of an employee by 4%.

In [ ]:
class Employee:
  def __init__(self, firstName, lastName, email, pay):
    self.firstName=firstName
    self.lastName=lastName
    self.email=email
    self.pay=pay

  def fullname(self):
    return emp_1.firstName+' '+emp_1.lastName

  def apply_inrement(self):
    self.pay = int(self.pay * 1.04)

In [ ]:
emp_1=Employee('FirstName_1', 'LastName_1','FirstName_1.LastName_1@email.com', 50000)
print(emp_1.pay)
emp_1.apply_inrement()
print(emp_1.pay)

50000
52000


In this code, the increment is not accessed from the code outside the class and also the same of all the employee. So, this is a good candidate for the class valriable.

In [ ]:
class Employee:
  annual_increment=1.1
  def __init__(self, firstName, lastName, email, pay):
    self.firstName=firstName
    self.lastName=lastName
    self.email=email
    self.pay=pay

  def fullname(self):
    return emp_1.firstName+' '+ emp_1.lastName

  def apply_inrement(self):
    self.pay = int(self.pay * self.annual_increment)
    # self.pay = int(self.pay * Employee.annual_increment)

We can access class variable using class itself or by instances. We can also check the attributes of the instances and the class using '__dict__()' method.

In [ ]:
emp_1=Employee('FirstName_1', 'LastName_1','FirstName_1.LastName_1@email.com', 50000)
emp_2=Employee('FirstName_2', 'LastName_2','FirstName_2.LastName_2@email.com', 50000)

In [ ]:
print(emp_1.pay)
emp_1.apply_inrement()
print(emp_1.pay)
print(emp_1.__dict__)
print(Employee.__dict__)

We can see that *annual_increment* appears as the attribute of the class but does not appear in the class instances (*emp_1*). We can change the value of the *annual_inrement* using the class. This will affect the attributes of the instances as well.

In [ ]:
Employee.annual_increment=1.2
print(Employee.annual_increment)
print(emp_1.annual_increment)
print(emp_2.annual_increment)

If we wish to change the class variable *annual_inrement* for a particular instance of the class, we can do so by assigning the value of the class variable *annual_increment* on an instance.

In [ ]:
emp_1.annual_increment=1.0
print(Employee.annual_increment)
emp_1.apply_inrement()
print(emp_1.pay)
emp_2.apply_inrement()
print(emp_2.pay)

1.1
50000
60500


This shows that annual increment is updated for the class variable and the annual increment is set a different value only for 'emp_1' instance.

In [ ]:
print(emp_1.__dict__)
print(emp_2.__dict__)

{'firstName': 'FirstName_1', 'lastName': 'LastName_1', 'email': 'FirstName_1.LastName_1@email.com', 'pay': 55000, 'annual_increment': 1.05}
{'firstName': 'FirstName_2', 'lastName': 'LastName_2', 'email': 'FirstName_2.LastName_2@email.com', 'pay': 60000}


So, setting *emp_1.annual_increment=1.05* added the class variable as the instance variable for *emp_1* instance. This concept become helpful to update the class variable for an instance separately. This is only possible when we access the class variable in a method by 'self' not by class.

In [ ]:
print(emp_1.pay)
emp_1.apply_inrement()
print(emp_1.pay)

57750
60637


Now, let us track the total number of employees, then, as it is the same for all the instance so is the good candidate for the class variable.

In [ ]:
class Employee:
  annual_increment=1.1
  numberEmployees=0
  def __init__(self, firstName, lastName, email, pay):
    self.firstName=firstName
    self.lastName=lastName
    self.email=email
    self.pay=pay
    Employee.numberEmployees += 1

  def fullname(self):
    return emp_1.firstName+' '+emp_1.lastName

  def apply_inrement(self):
    self.pay = int(self.pay * self.annual_increment)
    #self.pay = int(self.pay * Employee.annual_increment)

In [ ]:
emp_1=Employee('FirstName_1', 'LastName_1','FirstName_1.LastName_1@email.com', 50000)
emp_2=Employee('FirstName_2', 'LastName_2','FirstName_2.LastName_2@email.com', 60000)


In [ ]:
print(Employee.numberEmployees)

###Function overloading
The assignment of more than one behavior to a particular function. The operation performed varies by the types of objects or arguments involved.


###Inheritance
The transfer of the characteristics of a class to other classes that are derived from it.


###Operator overloading
The assignment of more than one function to a particular operator.

In [ ]:
class MyClass:

  # Regular method takes first arguement as instance itself
  # By convention we use 'self'

  def method(self):
    return 'instance method called', self

  @classmethod
  def classmethod(cls):
    return 'class method called', cls

  @staticmethod
  def staticmethod():
    return 'static method called'

##Instance Methods
The first method on MyClass, called method, is a regular instance method. That’s the basic method we use most of the time. We can see the method takes one parameter, self, which points to an instance of MyClass when the method is called (but of course instance methods can accept more than just one parameter).

Through the self parameter, instance methods can freely access attributes and other methods on the same object. This gives them a lot of power when it comes to modifying an object’s state.

Not only can they modify object state, instance methods can also access the class itself through the self.__class__ attribute. This means instance methods can also modify class state.

##Class Methods
Let’s compare that to the second method, MyClass.classmethod. We mark this method with a @classmethod decorator to flag it as a class method. Instead of accepting a self parameter, class methods take a cls parameter that points to the class—and not the object instance—when the method is called.

Because the class method only has access to this cls argument, it can’t modify object instance state. That would require access to self. However, class methods can still modify class state that applies across all instances of the class.

##Static Methods
The third method, MyClass.staticmethod is marked with a @staticmethod decorator to flag it as a static method.

This type of method takes neither a self nor a cls parameter (but of course it’s free to accept an arbitrary number of other parameters).

Therefore a static method can neither modify object state nor class state. Static methods are restricted in what data they can access - and they’re primarily a way to namespace your methods.

We’ll start by creating an instance of the class and then calling the three different methods on it.

MyClass was set up in such a way that each method’s implementation returns a tuple containing information for us to trace what’s going on — and which parts of the class or object the method can access.

Here’s what happens when we call an instance method:

In [ ]:
obj = MyClass()
obj.method()

('instance method called', <__main__.MyClass at 0x7f84705be160>)

This confirmed that method (the instance method) has access to the object instance (printed as <MyClass instance>) via the self argument.

In [ ]:
obj.classmethod()

('class method called', __main__.MyClass)

In [ ]:
MyClass.classmethod()

('class method called', __main__.MyClass)

In [ ]:
MyClass.staticmethod()

'static method called'

In [ ]:
MyClass.method()

TypeError: ignored

Will throw error as instance method need object to pass.

In [ ]:
MyClass.method(obj)

('instance method called', <__main__.MyClass at 0x7fed892ed070>)

Class methods are declared by using @classmethod and it takes class as argument.A class method can access or modify the class state.

In [ ]:
class Employee:
  increment = 1.5

  def __init__(self, fn, ln, sal):
      self.fname=fn
      self.lname=ln
      self.salary=sal

  def increase(self):
    self.salary = int(self.salary * self.increment)

  @classmethod
  def change_sal(cls, amount):
    cls.increment = amount

karan = Employee('karan', 'bhardwaj',50000)
karan2 = Employee('karan', 'bhardwaj',80000)


print(karan.salary)
Employee.change_sal(2)
karan.increase()
print(karan.salary)

50000
100000


**Class methods as alternative constructor**

In [ ]:
class Employee:
  increment = 1.5

  def __init__(self, fn, ln, sal):
      self.fname=fn
      self.lname=ln
      self.salary=sal

  def increase(self):
    self.salary = int(self.salary * self.increment)

  @classmethod
  def change_sal(cls, amount):
    cls.increment = amount

  @classmethod
  def

sachin = Employee('sachin','bharwal',80000)

print(sachin.salary)
Employee.change_sal(5)
sachin.increase()
print(sachin.salary)

80000
400000
